### ISSUE TITLE
Truncated JPEG loaded without error with "fail" flag

### ISSUE THREAD

**Issue by bozaro** [libvips/libvips#1946]
**Describe the bug**
I create small truncated jpeg for testing purpose and try to load it with "fail" = TRUE flag.
I excect error on loading image, but it loaded successfully. Warning shown only on libvips shutdown.

**To Reproduce**
Source code way:
1. Unpack [example.zip](https://github.com/libvips/libvips/files/5748004/example.zip)
2. Compile test tool: ```gcc read.c `pkg-config vips --cflags --libs` ```
3. Execute compiled tool: `./a.out broken.jpeg`

Binary way:
1. Unpack [example.zip](https://github.com/libvips/libvips/files/5748004/example.zip)
2. Execute command: `vips copy 'broken.jpeg[fail]' copy.jpeg`

**Expected behavior**
I expect error like: `VipsJpeg: Premature end of JPEG file`

**Actual behavior**
A clear and concise description of what actually happen.

**Screenshots**
Succesfull execution with output like (exit code 0):
```
➜  example ./a.out broken.jpeg
avg = 116.76

(a.out:502655): VIPS-WARNING **: 19:28:03.987: read gave 2 warnings

(a.out:502655): VIPS-WARNING **: 19:28:03.987: VipsJpeg: Premature end of JPEG file

```

**Environment**
 - OS: [Ubuntu 20.04.1 LTS]
 - Vips versions:
   - precompiled: vips-8.10.2-Sun Oct 18 19:45:39 UTC 2020
   - from source: aa6046d72827110c1dd6decffc49e5d5c6372481 (actual master)

**Additional context**
`broken.jpeg` and `broken.png` files was produced by removing last 16 bytes from `original.jpeg` and `original.png`.

**read.c**:
```
/* compile with
 * gcc read.c `pkg-config vips --cflags --libs`
 */

#include <vips/vips.h>
#include <stdio.h>

int
main( int argc, char **argv )
{
    VipsImage *im;
    double d;

    VIPS_INIT( argv[0] );

    if( !(im = vips_image_new_from_file( argv[1], "fail", TRUE, NULL )) )
        vips_error_exit( NULL ); 
    if( vips_avg( im, &d, NULL ) ) {
        g_object_unref( im );
        vips_error_exit( NULL ); 
    }
    g_object_unref( im );

    printf( "avg = %g\n", d ); 

    return( 0 );
}
```

**Comment by jcupitt** [comment]
Hello @bozaro,

This seems to be working for me with 8.10.5 (current stable). I see:

```
$ vips --version
vips-8.10.6-Mon Dec 28 18:26:19 UTC 2020
$ head -c -16 ~/pics/k2.jpg > x.jpg
$ ./a.out x.jpg

(a.out:873484): VIPS-WARNING **: 18:23:36.126: error in tile 0 x 2032

(a.out:873484): VIPS-WARNING **: 18:23:36.126: error in tile 0 x 2040
VipsJpeg: Premature end of JPEG file
VipsJpeg: out of order read at line 2032

$ 
```

**Comment by jcupitt** [comment]
I tried your `example.zip` and you're right, that one behaves differently:

```
$ ./a.out broken.jpeg 
avg = 116.76

(a.out:894239): VIPS-WARNING **: 18:34:36.415: read gave 2 warnings

(a.out:894239): VIPS-WARNING **: 18:34:36.415: VipsJpeg: Premature end of JPEG file

$
```

I think the problem is that libvips is a lazy library, so errors can be discovered very late. In this case, you have damaged the JPEG coda, the few bytes that end the file, and these are not read during pixel decode, but only during decompress shutdown.

For example:

```
$ head -c -100 original.jpeg > x.jpg
$ ./a.out x.jpg

(a.out:894411): VIPS-WARNING **: 18:38:18.530: error in tile 0 x 88
VipsJpeg: Premature end of JPEG file
```

Gives your desired result.

I think I'd class this as expected behaviour.

**Comment by bozaro** [comment]
I try to execute:

vips copy 'broken.jpeg[fail]' copy.png

This command creates copy.png file, show warning and exits with exit code 0.

I want to detect truncated image on loading or encoding stage in my code. And I cant use warning messages for this purpose in multithread image processing.

**Comment by jcupitt** [comment]
I looked through the code again, and you're right, the jpeg loader was not always signalling an error on EOF. It looks like this was a regression introduced in the recent switch to generic input objects.

I think I've fixed this in 8.10 HEAD, and credited you in the changelog. I now see:

```
$ ./a.out broken.jpeg 

(a.out:1124759): VIPS-WARNING **: 22:49:42.357: error in tile 0 x 88
VipsJpeg: Premature end of input file
$ echo $?
1
```

Thanks for reporting this!
